src/backend/distributed/master/master_repair_shards.c

```diff
@@ -457,12 +423,89 @@ ReplicateColocatedShardPlacement(int64 shardId, char *sourceNodeName,
 		EnsureReferenceTablesExistOnAllNodes();
 	}
 
+	CopyShardTables(colocatedShardList, sourceNodeName, sourceNodePort,
+					targetNodeName, targetNodePort);
+
 	/*
-	 * CopyColocatedShardPlacement function copies given shard with its co-located
-	 * shards.
+	 * Finally insert the placements to pg_dist_placement and sync it to the
+	 * metadata workers.
 	 */
-	CopyColocatedShardPlacement(shardId, sourceNodeName, sourceNodePort,
-								targetNodeName, targetNodePort);
+	foreach_ptr(colocatedShard, colocatedShardList)
+	{
+		uint64 colocatedShardId = colocatedShard->shardId;
+		uint32 groupId = GroupForNode(targetNodeName, targetNodePort);
+		uint64 placementId = GetNextPlacementId();
+
+		InsertShardPlacementRow(colocatedShardId, placementId,
+								SHARD_STATE_ACTIVE, ShardLength(colocatedShardId),
+								groupId);
+
+		if (ShouldSyncTableMetadata(colocatedShard->relationId))
+		{
+			char *placementCommand = PlacementUpsertCommand(colocatedShardId, placementId,
+															SHARD_STATE_ACTIVE, 0,
+															groupId);
+
+			SendCommandToWorkersWithMetadata(placementCommand);
+		}
+	}
+}
+
+
+/*
+ * EnsureTableListOwner ensures current user owns given tables. Superusers
+ * are regarded as owners.
+ */
+static void
+EnsureTableListOwner(List *tableIdList)
+{
+	Oid tableId = InvalidOid;
+	foreach_oid(tableId, tableIdList)
+	{
+		EnsureTableOwner(tableId);
```

Comment:
We should probably find a better way to do this, because it's quite possible that tables are co-located with different owners (perhaps we should prevent that). However, that's not this PR's fault and master_move_shard_placement has the same problem.